fix(SPEC): update SPEC and fix some inconsistencies (#5206)
influxdata/flux: docs/SPEC.md

The following keywords are reserved and may not be used as identifiers:

    and import  option   if
    or  package builtin  then
    not return  testcase else exists

#### Operators

The following character sequences represent operators:

    +   ==   !=   (   )   =>
    -   <    !~   [   ]   ^
    *   >    =~   {   }   ?

>     /   <=   =    ,   :   "

Reviewer: `?` and `"` are operators?

Author: The `?` shows up in type signatures for optional arguments. The `"` shows up for string literals. They are not operators in the sense of binary operators but they are still operators for the language. This list was taken from the scanner.rl implementation.

    +   ==   !=   (   )   =>
    -   <    !~   [   ]   ^
    *   >    =~   {   }
    /   <=   =    ,   :
    *   >    =~   {   }   ?
    /   <=   =    ,   :   "
    %   >=   <-   .   |>

> #### Numeric literals

Author: We tried to implement numeric literals but were unsuccessful in making it complete through the entire implementation. We can revisit this later but for now its no longer part of the SPEC.

Reviewer: `?` and `"` are operators?

    date.add(d: 1mo30d, to: 2018-01-01T00:00:00Z)                    // 2018-03-03T00:00:00Z, Months are applied first to get February 1st, then days are added resulting in March 3 in 2018.
    date.add(d: 1mo1d, to: 2018-01-31T00:00:00Z)                     // 2018-03-01T00:00:00Z, Months are applied first to get February 28th, then days are added resulting in March 1 in 2018.

    // Multiplication and addition of durations to date times
    date.add(d: date.scale(d:1mo, n:1), to: 2018-01-01T00:00:00Z)  // 2018-02-01T00:00:00Z
    date.add(d: date.scale(d:1mo, n:2), to: 2018-01-01T00:00:00Z)  // 2018-03-01T00:00:00Z
    date.add(d: date.scale(d:1mo, n:3), to: 2018-01-01T00:00:00Z)  // 2018-04-01T00:00:00Z
    date.add(d: date.scale(d:1mo, n:1), to: 2018-01-31T00:00:00Z)  // 2018-02-28T00:00:00Z
    date.add(d: date.scale(d:1mo, n:2), to: 2018-01-31T00:00:00Z)  // 2018-03-31T00:00:00Z
    date.add(d: date.scale(d:1mo, n:3), to: 2018-01-31T00:00:00Z)  // 2018-04-30T00:00:00Z

#### Date and time literals

A date and time literal represents a specific moment in time.
It has a date part, a time part and a time offset part.
The format follows the RFC 3339 specification.

> The time is optional, when it is omitted the time is assumed to be midnight UTC.

Reviewer 2: ```suggestion
The time is optional. When it is omitted the time is assumed to be midnight UTC.
```

> The time is optional, when it is omitted the time is assumed to be midnight UTC.

Author: The actual behavior implemented is to use UTC midnight, and no one is asking for the default location behavior as such the SPEC has been updated to reflect reality.

    date_time_lit     = date [ "T" time ] .
    date              = year_lit "-" month "-" day .
    year              = decimal_digit decimal_digit decimal_digit decimal_digit .
    month             = decimal_digit decimal_digit .
    day               = decimal_digit decimal_digit .

>     time              = hour ":" minute ":" second [ fractional_second ] time_offset .

Author: I left the implementation unchanged in that the time offset is optional so that its easier to provide a good error message to users if it is missing.

Reviewer: I'm not sure I understand. The way I read this change to the grammar, the `time_offset` was optional before, but now it is required.

Author: Correct now it is required, however the implementation will parse without a time offset and report the error that its missing.

Within the forward slashes, any unicode character may appear except for an unescaped forward slash.
The `\x` hex byte value representation from string literals may also be present.

> In addition standard escape sequences regular expression literals also support the following escape sequences:

Reviewer 2: ```suggestion
In addition standard escape sequences, regular expression literals also support the following escape sequences:
```

In addition standard escape sequences regular expression literals also support the following escape sequences:

    \/   U+002f forward slash
    \\   U+005c backslash

    regexp_lit         = "/" regexp_char { regexp_char } "/" .
    regexp_char        = unicode_char | byte_value | regexp_escape_char .
    regexp_escape_char = `\` (`/` | `\`)
    regexp_escape_char = `\/`

Examples:

    /.*/
    /http:\/\/localhost:8086/
    /^\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e(ZZ)?$/
    /^日本語(ZZ)?$/ // the above two lines are equivalent
    /\\xZZ/ // this becomes the literal pattern "\xZZ"

>     /a\/b\\c\d/ // escape sequences and character class shortcuts are supported

Author: There is no need to escape backslash in regex and it was never implemented, The SPEC has been updated to reflect that.

The string type is nullable.
Note that an empty string is distinct from a _null_ value.

> The length of a string is its size in bytes, not the number of characters, since a single character may be multiple bytes.

Author: Nothing in the SPEC depends on the length of a string as such this is best left documented in the stdlib strings package.

##### Function types

A _function type_ represents a set of all functions with the same argument and result types.

> Functions arguments are always named, meaning there are not positional arguments.

Reviewer: ```suggestion
Functions arguments are always named. (There are no positional arguments.)
```

> Functions arguments are always named, meaning there are not positional arguments.

Reviewer 2: ```suggestion
Functions arguments are always named. There are no positional arguments.
```

Therefore implementing a function type requires that the arguments be named the same.

##### Stream types

A _stream type_ represents an unbounded collection of values.

> The values may be of any other type but must all be the same type.

Reviewer: What does "any other" mean here? Does it mean there can't be nested stream types?

> The values may be of any other type but must all be the same type.

Reviewer 3: The values of a stream can't be any type, they must be records and those records may only hold int, uint, float, string, time or bool.

cc https://github.com/influxdata/flux/issues/5158

##### Function types

A _function type_ represents a set of all functions with the same argument and result types.
Functions arguments are always named. (There are no positional arguments.)
Therefore implementing a function type requires that the arguments be named the same.

##### Stream types

A _stream type_ represents an unbounded collection of values.

> The values must be records and those records may only hold int, uint, float, string, time or bool.

Reviewer: ```suggestion
The values must be records and those records may only hold int, uint, float, string, time or bool types.
```

Function literals are _closures_: they may refer to variables defined is a surrounding block.
Those variables are shared between the function literal and the surrounding block.

> Function arguments are named, there are no positional arguments.

Reviewer: ```suggestion
Function arguments are named. (There are no positional arguments.)
```
or
```suggestion
Function arguments are named.
```

Might not need to repeat this.

> Function arguments are named, there are no positional arguments.

Reviewer 2: ```suggestion
Function arguments are named. There are no positional arguments.
```

    BuiltinStatement = "builtin" identifier ":" TypeExpression .
    TypeExpression   = MonoType ["where" Constraints] .

    MonoType     = Tvar | BasicType | ArrayType | StreamType | VectorType | RecordType | FunctionType .
    Tvar         = "A" … "Z" .
    BasicType    = "int" | "uint" | "float" | "string" | "bool" | "time" | "duration" | "bytes" | "regexp" .
    ArrayType    = "[" MonoType "]" .

>     StreamType    = "stream" "[" MonoType "]" .
>     VectorType    = "vector" "[" MonoType "]" .

Reviewer: What about:
```suggestion
    StreamType    = "stream" ArrayType .
    VectorType    = "vector" ArrayType .
```

?

Author: I think that conflates the concepts of Stream and vector types. StreamType is not a stream of arrays rather it is its own thing. Same for vector types. So while its an equivalent grammar the semantics are a bit confusing using ArrayType.

Reviewer 4: We split Array and Stream purposefully since Stream represents a collection without a known size. Ref: https://github.com/influxdata/flux/issues/4344

    MonoType     = Tvar | BasicType | ArrayType | RecordType | FunctionType .
    Tvar         = "A" … "Z" .
    BasicType    = "int" | "uint" | "float" | "string" | "bool" | "time" | "duration" | "bytes" | "regexp" .

>     ArrayType    = [ "stream | "vector" ] "[" MonoType "]" .

Reviewer 3: I'd move the stream and vector to their own definition

```
StreamType = "stream" "[" MonoType "]"
// etc
```

or at least rename `ArrayType` to `CollectionType`

Example

>     builtin filter : (<-tables: stream[T], fn: (r: T) => bool) => stream[T]

Reviewer: Is `stream` a keyword or reserved string here? I don't see this added to the grammar anywhere yet.

Reviewer 2: `stream[]` is a type expression that is noted later in the SPEC.

Reviewer: Yes, but it should be part of the BNF grammar as well.

Author: Nice catch, added to the grammar, now. However its done in such a way that it doesn't need to be a keyword.

```

### Standard Library

> Flux provides a standard library of functions. Find documentation here https://docs.influxdata.com/flux/v0.x/stdlib/

Reviewer 2: Use the `latest` redirect to keep this current.

```suggestion
Flux provides a standard library of functions. Find documentation here https://docs.influxdata.com/flux/latest/stdlib/
```

##### replaceAllString

Returns a copy of v, replacing matches of the Regexp r with the replacement string t.

Example: `replaceAllString(r: regexp.compile("a(x*)b"), v: "-ab-axxb-", t: "T")` returns string `-T-T-`.

##### quoteMeta

Return a string that escapes all regular expression metacharacters inside the argument text; the returned string is a regular expression matching the literal text.

Example: `quoteMeta("Escaping symbols like: .+*?()|[]{}^$")` returns string `Escaping symbols like: \.\+\*\?\(\)\|\[\]\{\}\^\$`.

##### splitRegexp

Slices v into substrings separated by the expression and returns a slice of the substrings between those expression matches.

Example: `splitRegex(r: regexp.compile("a*"), v: "abaabaccadaaae", i: 5)` returns string array `["", "b", "b", "c", "cadaaae"]`.

#### JSON Operations

##### encode

Convert a value into JSON bytes.

Example: `encode(v: {label: "abcd})` returns JSON encoded value of the record as bytes.

Types `record` and `dictionary` are encoded as JSON objects.

### Composite data types
### Composite data types

A composite data type is a collection of primitive data types that together have a higher meaning.

> ### Triggers

Author: This concept was never exposed to users nor is it important to communicate.
